workitem/link/type_repository_blackbox_test.go

```diff
@@ -0,0 +1,318 @@
+		// given
+		fxt := tf.NewTestFixture(t, s.DB, tf.WorkItemLinkTypes(1))
+		typ := *fxt.WorkItemLinkTypes[0]
+		typ.ID = uuid.NewV4()
+		// when
+		createdType, err := s.typeRepo.Create(s.Ctx, &typ)
+		// then
+		require.Error(t, err)
+		require.IsType(t, errors.DataConflictError{}, errs.Cause(err))
+		require.Nil(t, createdType)
+	})
+}
+
+func (s *typeRepoBlackBoxTest) TestCheckExists() {
+	s.T().Run("existing", func(t *testing.T) {
+		// given
+		fxt := tf.NewTestFixture(t, s.DB, tf.WorkItemLinkTypes(1))
+		// when
+		err := s.typeRepo.CheckExists(s.Ctx, fxt.WorkItemLinkTypes[0].ID)
+		// then
+		require.NoError(t, err)
+	})
+	s.T().Run("nonexisting", func(t *testing.T) {
+		// given
+		linkTypeID := uuid.NewV4()
+		// when
+		err := s.typeRepo.CheckExists(s.Ctx, linkTypeID)
+		// then
+		require.Error(t, err)
+	})
+}
+
+func (s *typeRepoBlackBoxTest) TestSave() {
+	s.T().Run("version conflict", func(t *testing.T) {
+		// given
+		fxt := tf.NewTestFixture(t, s.DB, tf.WorkItemLinkTypes(1))
+		modelToSave := *fxt.WorkItemLinkTypes[0]
+		modelToSave.Version = modelToSave.Version + 10
+		// when
+		savedModel, err := s.typeRepo.Save(s.Ctx, modelToSave)
+		// then
+		require.Error(t, err)
+		require.IsType(t, errors.VersionConflictError{}, errs.Cause(err))
+		require.Nil(t, savedModel)
+	})
+	s.T().Run("space template reference changed (forbidden)", func(t *testing.T) {
+		// given
+		fxt := tf.NewTestFixture(t, s.DB, tf.WorkItemLinkTypes(1), tf.SpaceTemplates(2))
+		modelToSave := *fxt.WorkItemLinkTypes[0]
+		modelToSave.SpaceTemplateID = fxt.SpaceTemplates[1].ID
+		// when
+		savedModel, err := s.typeRepo.Save(s.Ctx, modelToSave)
+		// then
+		require.Error(t, err)
+		require.IsType(t, errors.ForbiddenError{}, errs.Cause(err))
+		require.Nil(t, savedModel)
+	})
+	s.T().Run("link type not found", func(t *testing.T) {
+		// given
+		fxt := tf.NewTestFixture(t, s.DB, tf.SpaceTemplates(1))
+		id := uuid.NewV4()
+		modelToSave := link.WorkItemLinkType{
+			ID:              id,
+			Name:            id.String(),
+			Description:     ptr.String("description for WILT " + id.String()),
+			ReverseName:     "reverse name",
+			ForwardName:     "forward name",
+			Topology:        link.TopologyTree,
+			LinkCategoryID:  link.SystemWorkItemLinkCategoryUserID,
+			SpaceTemplateID: fxt.SpaceTemplates[0].ID,
+		}
+		// when
+		savedModel, err := s.typeRepo.Save(s.Ctx, modelToSave)
+		// then
+		require.Error(t, err)
+		require.IsType(t, errors.NotFoundError{}, errs.Cause(err))
+		require.Nil(t, savedModel)
+	})
+	s.T().Run("unknown topology (internal error)", func(t *testing.T) {
+		// given
+		fxt := tf.NewTestFixture(t, s.DB, tf.WorkItemLinkTypes(1))
+		modelToSave := *fxt.WorkItemLinkTypes[0]
+		modelToSave.Topology = link.Topology("foobar")
+		// when
+		savedModel, err := s.typeRepo.Save(s.Ctx, modelToSave)
+		// then
+		require.Error(t, err)
+		require.IsType(t, errors.BadParameterError{}, errs.Cause(err))
+		require.Nil(t, savedModel)
+	})
+}
```

Comment:
might be redundant but we can have a test on `Save` for same name error.

Reply:
Will add it.

Reply:
Done in 75f261b0